helix-core/src/main/java/org/apache/helix/manager/zk/ZkBucketDataAccessor.java

```diff
@@ -0,0 +1,178 @@
+  private static final String BUCKET_SIZE_KEY = "BUCKET_SIZE";
+  private static final String DATA_SIZE_KEY = "DATA_SIZE";
+
+  // 100 KB for default bucket size
+  private static final int DEFAULT_BUCKET_SIZE = 100 * 1024;
+  private int _bucketSize = DEFAULT_BUCKET_SIZE;
+  private ZkSerializer _zkSerializer;
+  private BaseDataAccessor _zkBaseDataAccessor;
+
+  public ZkBucketDataAccessor(String zkAddr) {
+    HelixZkClient zkClient = DedicatedZkClientFactory.getInstance()
+        .buildZkClient(new HelixZkClient.ZkConnectionConfig(zkAddr));
+    zkClient.setZkSerializer(new DummySerializer());
+    _zkSerializer = new ZNRecordJacksonSerializer();
+    _zkBaseDataAccessor = new ZkBaseDataAccessor(zkClient);
+  }
+
+  @Override
+  public <T extends HelixProperty> boolean bucketWrite(String path, T value) {
+    throw new UnsupportedOperationException("bucketWrite not supported!");
+  }
+
+  @Override
+  public HelixProperty bucketRead(String path) {
+    throw new UnsupportedOperationException("bucketRead not supported!");
+  }
+
+  @Override
+  public <T extends HelixProperty> boolean compressedBucketWrite(String path, T value)
+      throws IOException {
+    // Take the ZNrecord and serialize it (get byte[])
+    byte[] serializedRecord = _zkSerializer.serialize(value.getRecord());
+
+    // Compress the byte[]
+    byte[] compressedRecord = GZipCompressionUtil.compress(serializedRecord);
+
+    // Compute N - number of buckets
+    int numBuckets = (compressedRecord.length + _bucketSize - 1) / _bucketSize;
+
+    List<String> paths = new ArrayList<>();
+    List<Object> buckets = new ArrayList<>();
+
+    // Add the metadata ZNode first
+    paths.add(path);
+    Map<String, Integer> metadata = new HashMap<>();
+    metadata.put(BUCKET_SIZE_KEY, _bucketSize);
+    metadata.put(DATA_SIZE_KEY, compressedRecord.length);
+    buckets.add(OBJECT_MAPPER.writeValueAsBytes(metadata));
+
+    int ptr = 0;
+    int counter = 0;
+    while (counter < numBuckets) {
+      paths.add(path + "/" + counter);
+      if (counter == numBuckets - 1) {
+        // Special treatment for the last bucket
+        buckets.add(
+            Arrays.copyOfRange(compressedRecord, ptr, ptr + compressedRecord.length % _bucketSize));
+      } else {
+        buckets.add(Arrays.copyOfRange(compressedRecord, ptr, ptr + _bucketSize));
+      }
+      ptr += _bucketSize;
+      counter++;
+    }
+
+    // Do an async set to ZK
+    boolean[] success;
+    success = _zkBaseDataAccessor.setChildren(paths, buckets, AccessOption.PERSISTENT);
+
+    // Return false if any of the writes failed
+    // TODO: Improve the failure handling
+    for (boolean s : success) {
+      if (!s) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @Override
+  public HelixProperty compressedBucketRead(String path) {
+    // Retrieve the metadata
+    Map metadataMap;
+    byte[] metadata = (byte[]) _zkBaseDataAccessor.get(path, null, AccessOption.PERSISTENT);
```

Comment:
Will it be easier if the metadata is a ZNode? So we will have the accessor to read directly. No need to "OBJECT_MAPPER.readValue()".

Reply:
I agree that serializing the metadata might be an overkill. Will fix.